src/basic/capability-util.c

```diff
@@ -138,6 +138,42 @@ int capability_ambient_set_apply(uint64_t set, bool also_inherit) {
         return 0;
 }
 
+int capability_ambient_set_clear(uint64_t set) {
+        _cleanup_cap_free_ cap_t tmp_cap = NULL;
+        unsigned long i;
+        int r;
+
+        tmp_cap = cap_get_proc();
+        if (!tmp_cap)
+                return -errno;
+
+        for (i = 0; i <= cap_last_cap(); i++) {
+                cap_value_t cap = (cap_value_t) i;
+
+                if (set & (UINT64_C(1) << i))
+                        /* Don't drop the capability if requested  */
+                        continue;
+
+                r = prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_IS_SET, i, 0, 0);
+                if (r < 0)
+                        return -errno;
+
+                if (r == 0)
+                        continue;
+
+                if (cap_set_flag(tmp_cap, CAP_INHERITABLE, 1, &cap, CAP_CLEAR) < 0)
+                        return -errno;
+
+                if (prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_LOWER, i, 0, 0) < 0)
+                        return -errno;
+        }
+
+        if (cap_set_proc(tmp_cap) < 0)
+                return -errno;
+
+        return 0;
+}
```

Comment:
shouldn't this be done in capability_ambient_set_apply()? i.e. that the function is changed to drop all ambient caps not specified and set all ambient caps specified? that would be much shorter and simpler and still do what you need, no?